File: plugins/system-report/skills/system-report/templates/reporter.cs
// =============================================================================
// system-report Reporter — TELEMETRY TUỲ CHỌN, MẶC ĐỊNH TẮT.
//
//   Gỡ bỏ hoàn toàn = xoá file này + 1 dòng wiring (`Report.Install();`). Hết.
//
// Bật:  REPORT_ENABLED=true  REPORT_INSTANCE={{INSTANCE}}  REPORT_REDIS_HOST=...
// Wiring (chọn ĐÚNG MỘT dòng):
//   · console/worker/service :  Report.Install();
//   · ASP.NET Core (DI)      :  builder.Logging.AddProvider(Report.Provider());
//     → nhánh này cần bật symbol MICROSOFT_EXTENSIONS_LOGGING (thêm vào .csproj:
//       <DefineConstants>$(DefineConstants);MICROSOFT_EXTENSIONS_LOGGING</DefineConstants>).
//       Không bật → dùng Report.Install() như trên.
//
// BẤT BIẾN (system-report §3.4):
//   · airtight  — nuốt MỌI exception; không throw ra caller, không block
//   · opt-in    — REPORT_ENABLED chưa bật thì Install()/Provider() là no-op tuyệt đối
//   · 1 file    — zero NuGet package, nói RESP thẳng qua TcpClient
//   · drop-safe — hàng đợi có trần; đầy thì VỨT log
//   · chỉ ghi namespace REPORT_* trên 1 Redis-log RIÊNG (không phải Redis production)
//
// Đổi namespace cho khớp dự án. Yêu cầu: .NET 6+.
// =============================================================================
using System;
using System.Collections.Concurrent;
using System.Globalization;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Ops.SystemReport
{
    public static class Report
    {
        private static readonly bool Enabled =
            (Environment.GetEnvironmentVariable("REPORT_ENABLED") ?? "false")
            .Trim().ToLowerInvariant() is "1" or "true" or "yes";

        private static readonly string Instance =
            Env("REPORT_INSTANCE", Environment.MachineName);
        private static readonly string Host = Env("REPORT_REDIS_HOST", "127.0.0.1");
        private static readonly int Port = EnvInt("REPORT_REDIS_PORT", 6399);
        private static readonly string Db = Env("REPORT_REDIS_DB", "0");
        private static readonly string Pass = Env("REPORT_REDIS_PASS", "");
        private static readonly int TtlSec = EnvInt("REPORT_TTL_DAYS", 7) * 86400;
        private static readonly int FlushSec = EnvInt("REPORT_FLUSH_SEC", 5);
        private static readonly int QueueMax = EnvInt("REPORT_QUEUE_MAX", 2000);

        private static readonly BlockingCollection<string> Queue =
            new BlockingCollection<string>(new ConcurrentQueue<string>(), Math.Max(16, QueueMax));

        private static int _started;
        private static long _dropped;

        private static string Env(string k, string d)
        {
            var v = Environment.GetEnvironmentVariable(k);
            return string.IsNullOrWhiteSpace(v) ? d : v;
        }

        private static int EnvInt(string k, int d) =>
            int.TryParse(Environment.GetEnvironmentVariable(k), out var n) ? n : d;

        /// <summary>Đẩy 1 dòng vào báo cáo ngày. Không bao giờ throw, không bao giờ block.</summary>
        public static void Log(string line)
        {
            if (!Enabled) return;
            try
            {
                var msg = DateTime.Now.ToString("HH:mm:ss", CultureInfo.InvariantCulture)
                          + " [" + Instance + "] " + (line ?? string.Empty).Replace("\r", " ");
                if (!Queue.TryAdd(msg)) Interlocked.Increment(ref _dropped);
            }
            catch { /* telemetry KHÔNG được làm phiền hệ thống */ }
        }

        /// <summary>1 dòng wiring duy nhất. Chưa bật REPORT_ENABLED → no-op tuyệt đối.</summary>
        public static void Install()
        {
            if (!Enabled) return;
            if (Interlocked.Exchange(ref _started, 1) == 1) return;
            try
            {
                Task.Factory.StartNew(PumpAsync, TaskCreationOptions.LongRunning);
                AppDomain.CurrentDomain.UnhandledException += (_, e) =>
                    Log("FATAL UnhandledException: " + (e.ExceptionObject as Exception)?.ToString());
                TaskScheduler.UnobservedTaskException += (_, e) =>
                {
                    Log("FATAL UnobservedTaskException: " + e.Exception);
                    e.SetObserved();
                };
                AppDomain.CurrentDomain.ProcessExit += (_, _) =>
                {
                    Log("instance stopping (dropped=" + Interlocked.Read(ref _dropped) + ")");
                    Queue.CompleteAdding();
                    Thread.Sleep(500); // cho worker 1 nhịp flush cuối; không chặn lâu
                };
                Log("instance started");
            }
            catch { /* airtight */ }
        }

        // ── RESP tối giản qua TcpClient ──────────────────────────────────────
        private static async Task PumpAsync()
        {
            TcpClient client = null;
            NetworkStream stream = null;
            var registered = false;
            var batch = new System.Collections.Generic.List<string>(256);

            while (true)
            {
                try
                {
                    batch.Clear();
                    if (!Queue.TryTake(out var first, Timeout.Infinite)) break; // CompleteAdding
                    batch.Add(first);
                    var deadline = Environment.TickCount64 + FlushSec * 1000L;
                    while (batch.Count < 200)
                    {
                        var left = (int)(deadline - Environment.TickCount64);
                        if (left <= 0 || !Queue.TryTake(out var more, left)) break;
                        batch.Add(more);
                    }

                    if (client == null || !client.Connected)
                    {
                        client?.Dispose();
                        client = new TcpClient { NoDelay = true };
                        await client.ConnectAsync(Host, Port).ConfigureAwait(false);
                        stream = client.GetStream();
                        registered = false;
                        if (Pass.Length > 0) await SendAsync(stream, "AUTH", Pass).ConfigureAwait(false);
                        if (Db != "0") await SendAsync(stream, "SELECT", Db).ConfigureAwait(false);
                        _ = Task.Run(() => DrainAsync(stream)); // nuốt reply
                    }

                    var key = "REPORT_DAILY:" + DateTime.Now.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)
                              + ":" + Instance;
                    if (!registered)
                    {
                        await SendAsync(stream, "SADD", "REPORT_INSTANCES", Instance).ConfigureAwait(false);
                        registered = true;
                    }
                    await SendAsync(stream, "APPEND", key, string.Join("\n", batch) + "\n").ConfigureAwait(false);
                    await SendAsync(stream, "EXPIRE", key, TtlSec.ToString(CultureInfo.InvariantCulture))
                        .ConfigureAwait(false);
                }
                catch (InvalidOperationException) { break; } // queue đã CompleteAdding + rỗng
                catch
                {
                    try { client?.Dispose(); } catch { }
                    client = null; stream = null; registered = false;
                    try { await Task.Delay(2000).ConfigureAwait(false); } catch { }
                }
            }
            try { client?.Dispose(); } catch { }
        }

        private static async Task SendAsync(NetworkStream s, params string[] args)
        {
            var sb = new StringBuilder().Append('*').Append(args.Length).Append("\r\n");
            foreach (var a in args)
                sb.Append('$').Append(Encoding.UTF8.GetByteCount(a)).Append("\r\n").Append(a).Append("\r\n");
            var bytes = Encoding.UTF8.GetBytes(sb.ToString());
            await s.WriteAsync(bytes, 0, bytes.Length).ConfigureAwait(false);
        }

        private static async Task DrainAsync(NetworkStream s)
        {
            var buf = new byte[4096];
            try { while (await s.ReadAsync(buf, 0, buf.Length).ConfigureAwait(false) > 0) { } }
            catch { /* ignore */ }
        }

        // ── Tuỳ chọn: ILoggerProvider cho ASP.NET Core ───────────────────────
        // Bỏ nguyên vùng #if nếu dự án không dùng Microsoft.Extensions.Logging.
#if MICROSOFT_EXTENSIONS_LOGGING
        public static Microsoft.Extensions.Logging.ILoggerProvider Provider()
        {
            Install();
            return new ReportLoggerProvider();
        }

        private sealed class ReportLoggerProvider : Microsoft.Extensions.Logging.ILoggerProvider
        {
            public Microsoft.Extensions.Logging.ILogger CreateLogger(string name) => new ReportLogger(name);
            public void Dispose() { }
        }

        private sealed class ReportLogger : Microsoft.Extensions.Logging.ILogger
        {
            private readonly string _name;
            public ReportLogger(string name) { _name = name; }
            public IDisposable BeginScope<TState>(TState state) => null;
            public bool IsEnabled(Microsoft.Extensions.Logging.LogLevel level) =>
                Enabled && level >= Microsoft.Extensions.Logging.LogLevel.Warning;
            public void Log<TState>(Microsoft.Extensions.Logging.LogLevel level,
                Microsoft.Extensions.Logging.EventId id, TState state, Exception ex,
                Func<TState, Exception, string> fmt)
            {
                if (!IsEnabled(level)) return;
                Report.Log(level + " " + _name + ": " + fmt(state, ex) + (ex != null ? " | " + ex : ""));
            }
        }
#else
        /// <summary>Stub khi chưa bật MICROSOFT_EXTENSIONS_LOGGING — dùng Install() thay thế.</summary>
        public static object Provider() { Install(); return null; }
#endif
    }
}

File: plugins/system-report/skills/system-report/templates/reporter.go
// =============================================================================
// system-report Reporter — TELEMETRY TUỲ CHỌN, MẶC ĐỊNH TẮT.
//
//	Gỡ bỏ hoàn toàn = xoá file này + 1 dòng wiring (`report.Install()`). Hết.
//
// Bật:  REPORT_ENABLED=true  REPORT_INSTANCE={{INSTANCE}}  REPORT_REDIS_HOST=...
// Wiring (đúng 1 dòng, đầu func main):
//
//	report.Install()
//
// BẤT BIẾN (system-report §3.4):
//   - airtight  — recover mọi panic; không bao giờ block caller
//   - opt-in    — REPORT_ENABLED chưa bật thì Install() là no-op tuyệt đối
//   - 1 file    — zero dependency (chỉ stdlib), nói RESP thẳng qua net.Conn
//   - drop-safe — channel có trần; đầy thì VỨT log
//   - chỉ ghi namespace REPORT_* trên 1 Redis-log RIÊNG (không phải Redis production)
//
// Đổi `package report` cho khớp layout dự án (vd đặt trong package main).
// =============================================================================
package report

import (
	"fmt"
	"log"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	enabled  = truthy(getenv("REPORT_ENABLED", "false"))
	instance = getenv("REPORT_INSTANCE", hostname())
	host     = getenv("REPORT_REDIS_HOST", "127.0.0.1")
	port     = getenv("REPORT_REDIS_PORT", "6399")
	db       = getenv("REPORT_REDIS_DB", "0")
	pass     = getenv("REPORT_REDIS_PASS", "")
	ttlDays  = atoi(getenv("REPORT_TTL_DAYS", "7"), 7)
	flushSec = atoi(getenv("REPORT_FLUSH_SEC", "5"), 5)
	queueMax = atoi(getenv("REPORT_QUEUE_MAX", "2000"), 2000)

	// cấp phát ngay lúc init: tránh data race giữa Log() và Install()
	ch      = makeCh()
	once    sync.Once
	dropped int64
	mu      sync.Mutex
)

func makeCh() chan string {
	if !enabled {
		return nil
	}
	return make(chan string, queueMax)
}

func getenv(k, d string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return d
}
func truthy(s string) bool {
	switch strings.ToLower(s) {
	case "1", "true", "yes":
		return true
	}
	return false
}
func atoi(s string, d int) int {
	if n, err := strconv.Atoi(s); err == nil {
		return n
	}
	return d
}
func hostname() string {
	h, err := os.Hostname()
	if err != nil {
		return "unknown"
	}
	return h
}

// ── RESP tối giản ────────────────────────────────────────────────────────────
func writeCmd(c net.Conn, args ...string) error {
	var b strings.Builder
	fmt.Fprintf(&b, "*%d\r\n", len(args))
	for _, a := range args {
		fmt.Fprintf(&b, "$%d\r\n%s\r\n", len(a), a)
	}
	_ = c.SetWriteDeadline(time.Now().Add(5 * time.Second))
	_, err := c.Write([]byte(b.String()))
	return err
}

func dial() (net.Conn, error) {
	c, err := net.DialTimeout("tcp", net.JoinHostPort(host, port), 5*time.Second)
	if err != nil {
		return nil, err
	}
	if pass != "" {
		if err := writeCmd(c, "AUTH", pass); err != nil {
			c.Close()
			return nil, err
		}
	}
	if db != "0" {
		if err := writeCmd(c, "SELECT", db); err != nil {
			c.Close()
			return nil, err
		}
	}
	go func() { // nuốt reply để kernel buffer không đầy
		buf := make([]byte, 4096)
		for {
			_ = c.SetReadDeadline(time.Now().Add(24 * time.Hour))
			if _, err := c.Read(buf); err != nil {
				return
			}
		}
	}()
	return c, nil
}

func worker() {
	defer func() { _ = recover() }()
	var conn net.Conn
	registered := false
	tick := time.NewTicker(time.Duration(flushSec) * time.Second)
	defer tick.Stop()
	batch := make([]string, 0, 200)

	send := func() {
		if len(batch) == 0 {
			return
		}
		defer func() { _ = recover() }()
		if conn == nil {
			c, err := dial()
			if err != nil {
				batch = batch[:0] // không giữ log cũ vô hạn
				return
			}
			conn, registered = c, false
		}
		key := "REPORT_DAILY:" + time.Now().Format("2006-01-02") + ":" + instance
		var err error
		if !registered {
			if err = writeCmd(conn, "SADD", "REPORT_INSTANCES", instance); err == nil {
				registered = true
			}
		}
		if err == nil {
			err = writeCmd(conn, "APPEND", key, strings.Join(batch, "\n")+"\n")
		}
		if err == nil {
			err = writeCmd(conn, "EXPIRE", key, strconv.Itoa(ttlDays*86400))
		}
		if err != nil {
			conn.Close()
			conn, registered = nil, false
		}
		batch = batch[:0]
	}

	for {
		select {
		case line, ok := <-ch:
			if !ok {
				send()
				return
			}
			batch = append(batch, line)
			if len(batch) >= 200 {
				send()
			}
		case <-tick.C:
			send()
		}
	}
}

// Log đẩy 1 dòng vào báo cáo ngày. Không bao giờ panic, không bao giờ block.
func Log(line string) {
	if !enabled || ch == nil {
		return
	}
	defer func() { _ = recover() }()
	msg := time.Now().Format("15:04:05") + " [" + instance + "] " + strings.ReplaceAll(line, "\r", " ")
	select {
	case ch <- msg:
	default: // hàng đợi đầy → vứt, KHÔNG chặn app
		mu.Lock()
		dropped++
		mu.Unlock()
	}
}

// Logf tiện cho call site có format.
func Logf(format string, a ...any) { Log(fmt.Sprintf(format, a...)) }

type teeWriter struct{ next *os.File }

func (t teeWriter) Write(p []byte) (int, error) {
	Log(strings.TrimRight(string(p), "\n"))
	return t.next.Write(p)
}

// Install là 1 dòng wiring duy nhất. Chưa bật REPORT_ENABLED → no-op tuyệt đối.
func Install() {
	if !enabled {
		return
	}
	once.Do(func() {
		defer func() { _ = recover() }()
		go worker()
		log.SetOutput(teeWriter{next: os.Stderr}) // mirror log chuẩn vào báo cáo
		Log("instance started")
	})
}
